### ISSUE TITLE
Local methods don't inherit unsafe context from type, method or block where they are declared

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#13172]
Test code included at the bottom of this report.

According to the latest version of specification if block, type or method (property, event, indexer, operator, constructor, destructor, static constructor etc) is declared as unsafe all its content is considered an unsafe context.

> An unsafe context is introduced by including an unsafe modifier in the declaration of a type or member, or by employing an unsafe-statement:
> 
> ⦁   A declaration of a class, struct, interface, or delegate may include an unsafe modifier, in which case the entire textual extent of that type declaration (including the body of the class, struct, or interface) is considered an unsafe context.
> ⦁   A declaration of a field, method, property, event, indexer, operator, instance constructor, destructor, or static constructor may include an unsafe modifier, in which case the entire textual extent of that member declaration is considered an unsafe context.
> ⦁   An unsafe-statement enables the use of an unsafe context within a block. The entire textual extent of the associated block is considered an unsafe context.

However local methods do not use these rules and always define their own safe context unless they are explicitly marked as unsafe. This behavior is not consistent with neither lambda expressions nor anonymous methods since both mentioned expression types work according to the specification. 

Marking each local method as unsafe explicitly in an unsafe block/method/type may be annoying. 

``` cs
using System;
using System.Threading.Tasks;
public unsafe struct C
{
    public unsafe void M(int* p)
    {
        Console.WriteLine(*p);
        unsafe
        {
            void Local()
            {
                Console.WriteLine(*p); // error - cannot use pointers in safe context
                await Task.Delay(1000); // await cannot be used in unsafe context but Local defined its own safe context
            }
        }
    }
}
```

**Comment by jaredpar** [comment]
Thanks for reporting.  Unsure at a glance if this is definitely a bug but it does seem so.  Will let @agocke chime in.

**Comment by agocke** [comment]
@TessenR @jaredpar I don't believe local functions have been explicitly specced either way. I don't see a problem with allowing it, though.

@MadsTorgersen Is there any opinion from the language side here?

**Comment by tmat** [comment]
I think it should be consistent with lambdas.

**Comment by TessenR** [comment]
@agocke It's not obvious what is intended behavior and what isn't without specification.
I've reported this because current behavior is bizarre and not consistent with lambdas and regular/anonymous methods.

I'd really appreciate any information about local methods' intended behavior in this case.

**Comment by agocke** [comment]
@TessenR I agree with @jaredpar and @tmat. I'm going to treat this as a bug for now.
